cmd/lotus-shed/indexes.go

```diff
@@ -483,87 +492,145 @@ var inspectEventsCmd = &cli.Command{
+			tsKeyCid, err := currTs.Key().Cid()
 			if err != nil {
-				return fmt.Errorf("failed to get tipset key cid: %w", err)
+				return xerrors.Errorf("failed to get tipset key cid: %w", err)
 			}
 
-			var expectEvents int
-			var expectEntries int
+			var problems []string
+			var hasEvents bool
 
-			for _, receipt := range receipts {
-				if receipt.ExitCode != exitcode.Ok || receipt.EventsRoot == nil {
-					continue
+			checkEventAndEntryCounts := func() error {
+				// compare by counting events, using ChainGetEvents to load the events from the chain
+				expectEvents, expectEntries, err := chainEventAndEntryCountsAt(ctx, currTs, receipts, api)
+				if err != nil {
+					return err
 				}
-				events, err := api.ChainGetEvents(ctx, *receipt.EventsRoot)
+				if expectEvents > 0 {
+					hasEvents = true
+				}
+
+				actualEvents, actualEntries, err := dbEventAndEntryCountsAt(currTs, stmtEventCount, stmtEntryCount)
 				if err != nil {
-					return fmt.Errorf("failed to load events for tipset %s: %w", currTs, err)
+					return err
 				}
-				expectEvents += len(events)
-				for _, event := range events {
-					expectEntries += len(event.Entries)
+
+				if actualEvents != expectEvents {
+					problems = append(problems, fmt.Sprintf("expected %d events, got %d", expectEvents, actualEvents))
+				}
+				if actualEntries != expectEntries {
+					problems = append(problems, fmt.Sprintf("expected %d entries, got %d", expectEntries, actualEntries))
 				}
+
+				return nil
 			}
 
-			var problems []string
+			// Compare the AMT roots: we reconstruct the event AMT from the database data we have and
+			// compare it with the on-chain AMT root from the receipt. If it's the same CID then we have
+			// exactly the same event data. Any variation, in number of events, and even a single byte
+			// in event data, will be considered a mismatch.
+
+			// cache for address -> actorID because it's typical for tipsets to generate many events for
+			// the same actors so we can try and avoid too many StateLookupID calls
+			addrIdCache := make(map[address.Address]abi.ActorID)
+
+			eventIndex := 0
+			for msgIndex, receipt := range receipts {
+				if receipt.EventsRoot == nil {
+					continue
+				}
+
+				amtRoot, has, problem, err := amtRootForEvents(
+					ctx,
+					api,
+					tsKeyCid,
+					prevTs.Key(),
+					stmtSelectEventIdAndEmitter,
+					stmtSelectEventEntries,
+					messages[msgIndex],
+					addrIdCache,
+				)
+				if err != nil {
+					return err
+				}
+				if has && !hasEvents {
+					hasEvents = true
+				}
+
+				if problem != "" {
+					problems = append(problems, problem)
+				} else if amtRoot != *receipt.EventsRoot {
+					problems = append(problems, fmt.Sprintf("events root mismatch for event %d", eventIndex))
+					// also provide more information about the mismatch
+					if err := checkEventAndEntryCounts(); err != nil {
```

Comment:
The `checkEventAndEntryCounts` only matches on counts. I think we should also log a "counts are same but roots dont match" or something similar to point out this specific problem to the user.

Reply:
Yeah, that's what will happen here - the reason I call `checkEventAndEntryCounts` is that the error for that tipset will be something like "✗ [1234567] (bafyfoo): events root mismatch for event 5, expected 10 events but got 14, expected 100 event entries but got 199". If the counts match then it'd just be "✗ [1234567] (bafyfoo): events root mismatch for event 5".

But looking at this error I realise it's a little bit wrong, it's not "event X" it's "message X" and "X" isn't really useful, so I'm replacing it with:

```go
problems = append(problems, fmt.Sprintf("events root mismatch for message %s", messages[msgIndex].Cid))
```